hugr-core/src/hugr/validate.rs

```diff
@@ -62,48 +56,37 @@ impl Hugr {
         validator.validate()
     }
 
-    /// Validate extensions on the input and output edges of nodes. Check that
-    /// the target ends of edges require the extensions from the sources, and
-    /// check extension deltas from parent nodes are reflected in their children.
-    pub fn validate_extensions(&self, closure: ExtensionSolution) -> Result<(), ValidationError> {
-        let validator = ExtensionValidator::new(self, closure);
-        for src_node in self.nodes() {
-            let node_type = self.get_nodetype(src_node);
-
-            // FuncDefns have no resources since they're static nodes, but the
-            // functions they define can have any extension delta.
-            if node_type.tag() != OpTag::FuncDefn {
-                // If this is a container with I/O nodes, check that the extension they
-                // define match the extensions of the container.
-                if let Some([input, output]) = self.get_io(src_node) {
-                    validator.validate_io_extensions(src_node, input, output)?;
-                }
-            }
-
-            for src_port in self.node_outputs(src_node) {
-                for (tgt_node, tgt_port) in self.linked_inputs(src_node, src_port) {
-                    validator.check_extensions_compatible(
-                        &(src_node, src_port.into()),
-                        &(tgt_node, tgt_port.into()),
-                    )?;
+    /// Validate extensions, i.e. that extension deltas from parent nodes are reflected in their children.
+    pub fn validate_extensions(&self) -> Result<(), ValidationError> {
+        for parent in self.nodes() {
+            let parent_op = self.get_optype(parent);
+            let parent_extensions = match parent_op.inner_function_type() {
+                Some(FunctionType { extension_reqs, .. }) => extension_reqs,
+                None => match parent_op.tag() {
+                    OpTag::Cfg | OpTag::Conditional => parent_op.extension_delta(),
```

Comment:
No, not any more - TailLoop is a DataflowParent like Dfg, FuncDefn and Case. Cfg and Conditional are the only ones here because they are not DataflowParents as their children are not dataflow nodes.

Reply:
(that is to say, `TailLoop` now has an `inner_function_type` so it gets handled two lines above)